// @aries-framework/core@0.1.0/build/modules/credentials/messages/OfferCredentialMessage.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OfferCredentialMessage = exports.INDY_CREDENTIAL_OFFER_ATTACHMENT_ID = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const AgentMessage_1 = require("../../../agent/AgentMessage");
const Attachment_1 = require("../../../decorators/attachment/Attachment");
const CredentialPreview_1 = require("./CredentialPreview");
exports.INDY_CREDENTIAL_OFFER_ATTACHMENT_ID = 'libindy-cred-offer-0';
/**
 * Message part of Issue Credential Protocol used to continue or initiate credential exchange by issuer.
 *
 * @see https://github.com/hyperledger/aries-rfcs/blob/master/features/0036-issue-credential/README.md#offer-credential
 */
class OfferCredentialMessage extends AgentMessage_1.AgentMessage {
    constructor(options) {
        super();
        this.type = OfferCredentialMessage.type;
        if (options) {
            this.id = options.id || this.generateId();
            this.comment = options.comment;
            this.credentialPreview = options.credentialPreview;
            this.offerAttachments = options.offerAttachments;
            this.attachments = options.attachments;
        }
    }
    get indyCredentialOffer() {
        var _a, _b;
        const attachment = this.offerAttachments.find((attachment) => attachment.id === exports.INDY_CREDENTIAL_OFFER_ATTACHMENT_ID);
        // Extract credential offer from attachment
        const credentialOfferJson = (_b = (_a = attachment === null || attachment === void 0 ? void 0 : attachment.data) === null || _a === void 0 ? void 0 : _a.getDataAsJson()) !== null && _b !== void 0 ? _b : null;
        return credentialOfferJson;
    }
}
OfferCredentialMessage.type = 'https://didcomm.org/issue-credential/1.0/offer-credential';
__decorate([
    class_validator_1.Equals(OfferCredentialMessage.type),
    __metadata("design:type", Object)
], OfferCredentialMessage.prototype, "type", void 0);
__decorate([
    class_validator_1.IsString(),
    class_validator_1.IsOptional(),
    __metadata("design:type", String)
], OfferCredentialMessage.prototype, "comment", void 0);
__decorate([
    class_transformer_1.Expose({ name: 'credential_preview' }),
    class_transformer_1.Type(() => CredentialPreview_1.CredentialPreview),
    class_validator_1.ValidateNested(),
    class_validator_1.IsInstance(CredentialPreview_1.CredentialPreview),
    __metadata("design:type", CredentialPreview_1.CredentialPreview)
], OfferCredentialMessage.prototype, "credentialPreview", void 0);
__decorate([
    class_transformer_1.Expose({ name: 'offers~attach' }),
    class_transformer_1.Type(() => Attachment_1.Attachment),
    class_validator_1.IsArray(),
    class_validator_1.ValidateNested({
        each: true,
    }),
    class_validator_1.IsInstance(Attachment_1.Attachment, { each: true }),
    __metadata("design:type", Array)
], OfferCredentialMessage.prototype, "offerAttachments", void 0);
exports.OfferCredentialMessage = OfferCredentialMessage;
