// @aries-framework/core@0.1.0/build/modules/credentials/messages/ProposeCredentialMessage.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProposeCredentialMessage = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const AgentMessage_1 = require("../../../agent/AgentMessage");
const CredentialPreview_1 = require("./CredentialPreview");
/**
 * Message part of Issue Credential Protocol used to initiate credential exchange by prover.
 *
 * @see https://github.com/hyperledger/aries-rfcs/blob/master/features/0036-issue-credential/README.md#propose-credential
 */
class ProposeCredentialMessage extends AgentMessage_1.AgentMessage {
    constructor(options) {
        var _a;
        super();
        this.type = ProposeCredentialMessage.type;
        if (options) {
            this.id = (_a = options.id) !== null && _a !== void 0 ? _a : this.generateId();
            this.comment = options.comment;
            this.credentialProposal = options.credentialProposal;
            this.schemaIssuerDid = options.schemaIssuerDid;
            this.schemaId = options.schemaId;
            this.schemaName = options.schemaName;
            this.schemaVersion = options.schemaVersion;
            this.credentialDefinitionId = options.credentialDefinitionId;
            this.issuerDid = options.issuerDid;
            this.attachments = options.attachments;
        }
    }
}
ProposeCredentialMessage.type = 'https://didcomm.org/issue-credential/1.0/propose-credential';
__decorate([
    class_validator_1.Equals(ProposeCredentialMessage.type),
    __metadata("design:type", Object)
], ProposeCredentialMessage.prototype, "type", void 0);
__decorate([
    class_validator_1.IsOptional(),
    class_validator_1.IsString(),
    __metadata("design:type", String)
], ProposeCredentialMessage.prototype, "comment", void 0);
__decorate([
    class_transformer_1.Expose({ name: 'credential_proposal' }),
    class_transformer_1.Type(() => CredentialPreview_1.CredentialPreview),
    class_validator_1.ValidateNested(),
    class_validator_1.IsOptional(),
    class_validator_1.IsInstance(CredentialPreview_1.CredentialPreview),
    __metadata("design:type", CredentialPreview_1.CredentialPreview
    /**
     * Filter to request credential based on a particular Schema issuer DID.
     */
    )
], ProposeCredentialMessage.prototype, "credentialProposal", void 0);
__decorate([
    class_transformer_1.Expose({ name: 'schema_issuer_did' }),
    class_validator_1.IsString(),
    class_validator_1.IsOptional(),
    __metadata("design:type", String)
], ProposeCredentialMessage.prototype, "schemaIssuerDid", void 0);
__decorate([
    class_transformer_1.Expose({ name: 'schema_id' }),
    class_validator_1.IsString(),
    class_validator_1.IsOptional(),
    __metadata("design:type", String)
], ProposeCredentialMessage.prototype, "schemaId", void 0);
__decorate([
    class_transformer_1.Expose({ name: 'schema_name' }),
    class_validator_1.IsString(),
    class_validator_1.IsOptional(),
    __metadata("design:type", String)
], ProposeCredentialMessage.prototype, "schemaName", void 0);
__decorate([
    class_transformer_1.Expose({ name: 'schema_version' }),
    class_validator_1.IsString(),
    class_validator_1.IsOptional(),
    __metadata("design:type", String)
], ProposeCredentialMessage.prototype, "schemaVersion", void 0);
__decorate([
    class_transformer_1.Expose({ name: 'cred_def_id' }),
    class_validator_1.IsString(),
    class_validator_1.IsOptional(),
    __metadata("design:type", String)
], ProposeCredentialMessage.prototype, "credentialDefinitionId", void 0);
__decorate([
    class_transformer_1.Expose({ name: 'issuer_did' }),
    class_validator_1.IsString(),
    class_validator_1.IsOptional(),
    __metadata("design:type", String)
], ProposeCredentialMessage.prototype, "issuerDid", void 0);
exports.ProposeCredentialMessage = ProposeCredentialMessage;
