// @aries-framework/core@0.1.0/build/modules/credentials/messages/RequestCredentialMessage.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestCredentialMessage = exports.INDY_CREDENTIAL_REQUEST_ATTACHMENT_ID = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const AgentMessage_1 = require("../../../agent/AgentMessage");
const Attachment_1 = require("../../../decorators/attachment/Attachment");
exports.INDY_CREDENTIAL_REQUEST_ATTACHMENT_ID = 'libindy-cred-request-0';
class RequestCredentialMessage extends AgentMessage_1.AgentMessage {
    constructor(options) {
        super();
        this.type = RequestCredentialMessage.type;
        if (options) {
            this.id = options.id || this.generateId();
            this.comment = options.comment;
            this.requestAttachments = options.requestAttachments;
            this.attachments = options.attachments;
        }
    }
    get indyCredentialRequest() {
        var _a, _b;
        const attachment = this.requestAttachments.find((attachment) => attachment.id === exports.INDY_CREDENTIAL_REQUEST_ATTACHMENT_ID);
        // Extract proof request from attachment
        const credentialReqJson = (_b = (_a = attachment === null || attachment === void 0 ? void 0 : attachment.data) === null || _a === void 0 ? void 0 : _a.getDataAsJson()) !== null && _b !== void 0 ? _b : null;
        return credentialReqJson;
    }
}
RequestCredentialMessage.type = 'https://didcomm.org/issue-credential/1.0/request-credential';
__decorate([
    class_validator_1.Equals(RequestCredentialMessage.type),
    __metadata("design:type", Object)
], RequestCredentialMessage.prototype, "type", void 0);
__decorate([
    class_validator_1.IsString(),
    class_validator_1.IsOptional(),
    __metadata("design:type", String)
], RequestCredentialMessage.prototype, "comment", void 0);
__decorate([
    class_transformer_1.Expose({ name: 'requests~attach' }),
    class_transformer_1.Type(() => Attachment_1.Attachment),
    class_validator_1.IsArray(),
    class_validator_1.ValidateNested({
        each: true,
    }),
    class_validator_1.IsInstance(Attachment_1.Attachment, { each: true }),
    __metadata("design:type", Array)
], RequestCredentialMessage.prototype, "requestAttachments", void 0);
exports.RequestCredentialMessage = RequestCredentialMessage;
