// @aries-framework/core@0.1.0/build/utils/BufferEncoder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BufferEncoder = void 0;
const base58_1 = require("./base58");
const base64_1 = require("./base64");
const buffer_1 = require("./buffer");
class BufferEncoder {
    /**
     * Encode buffer into base64 string.
     *
     * @param buffer the buffer to encode into base64 string
     */
    static toBase64(buffer) {
        return buffer_1.Buffer.from(buffer).toString('base64');
    }
    /**
     * Encode buffer into base64url string.
     *
     * @param buffer the buffer to encode into base64url string
     */
    static toBase64URL(buffer) {
        return base64_1.base64ToBase64URL(BufferEncoder.toBase64(buffer));
    }
    /**
     * Encode buffer into base58 string.
     *
     * @param buffer the buffer to encode into base58 string
     */
    static toBase58(buffer) {
        return base58_1.encodeToBase58(buffer);
    }
    /**
     * Decode base64 string into buffer. Also supports base64url
     *
     * @param base64 the base64 or base64url string to decode into buffer format
     */
    static fromBase64(base64) {
        return buffer_1.Buffer.from(base64, 'base64');
    }
    /**
     * Decode base58 string into buffer
     *
     * @param base58 the base58 string to decode into buffer format
     */
    static fromBase58(base58) {
        return buffer_1.Buffer.from(base58_1.decodeFromBase58(base58));
    }
    /**
     * Decode string into buffer.
     *
     * @param str the string to decode into buffer format
     */
    static fromString(str) {
        return buffer_1.Buffer.from(str);
    }
    static toUtf8String(buffer) {
        return buffer_1.Buffer.from(buffer).toString();
    }
}
exports.BufferEncoder = BufferEncoder;
