// @aries-framework/core@0.1.0/build/modules/credentials/repository/CredentialRecord.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CredentialRecord = void 0;
const class_transformer_1 = require("class-transformer");
const Attachment_1 = require("../../../decorators/attachment/Attachment");
const error_1 = require("../../../error");
const BaseRecord_1 = require("../../../storage/BaseRecord");
const uuid_1 = require("../../../utils/uuid");
const messages_1 = require("../messages");
const CredentialInfo_1 = require("../models/CredentialInfo");
class CredentialRecord extends BaseRecord_1.BaseRecord {
    constructor(props) {
        var _a, _b, _c;
        super();
        this.type = CredentialRecord.type;
        if (props) {
            this.id = (_a = props.id) !== null && _a !== void 0 ? _a : uuid_1.uuid();
            this.createdAt = (_b = props.createdAt) !== null && _b !== void 0 ? _b : new Date();
            this.state = props.state;
            this.connectionId = props.connectionId;
            this.credentialId = props.credentialId;
            this.threadId = props.threadId;
            this._tags = (_c = props.tags) !== null && _c !== void 0 ? _c : {};
            this.proposalMessage = props.proposalMessage;
            this.offerMessage = props.offerMessage;
            this.requestMessage = props.requestMessage;
            this.credentialMessage = props.credentialMessage;
            this.credentialAttributes = props.credentialAttributes;
            this.autoAcceptCredential = props.autoAcceptCredential;
            this.linkedAttachments = props.linkedAttachments;
            this.errorMessage = props.errorMessage;
        }
    }
    getTags() {
        return Object.assign(Object.assign({}, this._tags), { threadId: this.threadId, connectionId: this.connectionId, state: this.state, credentialId: this.credentialId });
    }
    getCredentialInfo() {
        if (!this.credentialAttributes)
            return null;
        const claims = this.credentialAttributes.reduce((accumulator, current) => (Object.assign(Object.assign({}, accumulator), { [current.name]: current.value })), {});
        return new CredentialInfo_1.CredentialInfo({
            claims,
            attachments: this.linkedAttachments,
            metadata: this.metadata.data,
        });
    }
    assertState(expectedStates) {
        if (!Array.isArray(expectedStates)) {
            expectedStates = [expectedStates];
        }
        if (!expectedStates.includes(this.state)) {
            throw new error_1.AriesFrameworkError(`Credential record is in invalid state ${this.state}. Valid states are: ${expectedStates.join(', ')}.`);
        }
    }
    assertConnection(currentConnectionId) {
        if (!this.connectionId) {
            throw new error_1.AriesFrameworkError(`Credential record is not associated with any connection. This is often the case with connection-less credential exchange`);
        }
        else if (this.connectionId !== currentConnectionId) {
            throw new error_1.AriesFrameworkError(`Credential record is associated with connection '${this.connectionId}'. Current connection is '${currentConnectionId}'`);
        }
    }
}
CredentialRecord.type = 'CredentialRecord';
__decorate([
    class_transformer_1.Type(() => messages_1.ProposeCredentialMessage),
    __metadata("design:type", messages_1.ProposeCredentialMessage)
], CredentialRecord.prototype, "proposalMessage", void 0);
__decorate([
    class_transformer_1.Type(() => messages_1.OfferCredentialMessage),
    __metadata("design:type", messages_1.OfferCredentialMessage)
], CredentialRecord.prototype, "offerMessage", void 0);
__decorate([
    class_transformer_1.Type(() => messages_1.RequestCredentialMessage),
    __metadata("design:type", messages_1.RequestCredentialMessage)
], CredentialRecord.prototype, "requestMessage", void 0);
__decorate([
    class_transformer_1.Type(() => messages_1.IssueCredentialMessage),
    __metadata("design:type", messages_1.IssueCredentialMessage)
], CredentialRecord.prototype, "credentialMessage", void 0);
__decorate([
    class_transformer_1.Type(() => messages_1.CredentialPreviewAttribute),
    __metadata("design:type", Array)
], CredentialRecord.prototype, "credentialAttributes", void 0);
__decorate([
    class_transformer_1.Type(() => Attachment_1.Attachment),
    __metadata("design:type", Array)
], CredentialRecord.prototype, "linkedAttachments", void 0);
exports.CredentialRecord = CredentialRecord;
