// @aries-framework/core@0.1.0/build/modules/indy/services/IndyIssuerService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndyIssuerService = void 0;
const tsyringe_1 = require("tsyringe");
const AgentConfig_1 = require("../../../agent/AgentConfig");
const AriesFrameworkError_1 = require("../../../error/AriesFrameworkError");
const IndySdkError_1 = require("../../../error/IndySdkError");
const indyError_1 = require("../../../utils/indyError");
const path_1 = require("../../../utils/path");
const IndyWallet_1 = require("../../../wallet/IndyWallet");
let IndyIssuerService = class IndyIssuerService {
    constructor(agentConfig, wallet) {
        this.indy = agentConfig.agentDependencies.indy;
        this.wallet = wallet;
        this.fileSystem = agentConfig.fileSystem;
    }
    /**
     * Create a new credential schema.
     *
     * @returns the schema.
     */
    async createSchema({ originDid, name, version, attributes }) {
        try {
            const [, schema] = await this.indy.issuerCreateSchema(originDid, name, version, attributes);
            return schema;
        }
        catch (error) {
            throw new IndySdkError_1.IndySdkError(error);
        }
    }
    /**
     * Create a new credential definition and store it in the wallet.
     *
     * @returns the credential definition.
     */
    async createCredentialDefinition({ issuerDid, schema, tag = 'default', signatureType = 'CL', supportRevocation = false, }) {
        try {
            const [, credentialDefinition] = await this.indy.issuerCreateAndStoreCredentialDef(this.wallet.handle, issuerDid, schema, tag, signatureType, {
                support_revocation: supportRevocation,
            });
            return credentialDefinition;
        }
        catch (error) {
            throw new IndySdkError_1.IndySdkError(error);
        }
    }
    /**
     * Create a credential offer for the given credential definition id.
     *
     * @param credentialDefinitionId The credential definition to create an offer for
     * @returns The created credential offer
     */
    async createCredentialOffer(credentialDefinitionId) {
        try {
            return await this.indy.issuerCreateCredentialOffer(this.wallet.handle, credentialDefinitionId);
        }
        catch (error) {
            throw new IndySdkError_1.IndySdkError(error);
        }
    }
    /**
     * Create a credential.
     *
     * @returns Credential and revocation id
     */
    async createCredential({ credentialOffer, credentialRequest, credentialValues, revocationRegistryId, tailsFilePath, }) {
        try {
            // Indy SDK requires tailsReaderHandle. Use null if no tailsFilePath is present
            const tailsReaderHandle = tailsFilePath ? await this.createTailsReader(tailsFilePath) : 0;
            if (revocationRegistryId || tailsFilePath) {
                throw new AriesFrameworkError_1.AriesFrameworkError('Revocation not supported yet');
            }
            const [credential, credentialRevocationId] = await this.indy.issuerCreateCredential(this.wallet.handle, credentialOffer, credentialRequest, credentialValues, revocationRegistryId !== null && revocationRegistryId !== void 0 ? revocationRegistryId : null, tailsReaderHandle);
            return [credential, credentialRevocationId];
        }
        catch (error) {
            if (indyError_1.isIndyError(error)) {
                throw new IndySdkError_1.IndySdkError(error);
            }
            throw error;
        }
    }
    /**
     * Get a handler for the blob storage tails file reader.
     *
     * @param tailsFilePath The path of the tails file
     * @returns The blob storage reader handle
     */
    async createTailsReader(tailsFilePath) {
        try {
            const tailsFileExists = await this.fileSystem.exists(tailsFilePath);
            // Extract directory from path (should also work with windows paths)
            const dirname = path_1.getDirFromFilePath(tailsFilePath);
            if (!tailsFileExists) {
                throw new AriesFrameworkError_1.AriesFrameworkError(`Tails file does not exist at path ${tailsFilePath}`);
            }
            const tailsReaderConfig = {
                base_dir: dirname,
            };
            return await this.indy.openBlobStorageReader('default', tailsReaderConfig);
        }
        catch (error) {
            if (indyError_1.isIndyError(error)) {
                throw new IndySdkError_1.IndySdkError(error);
            }
            throw error;
        }
    }
};
IndyIssuerService = __decorate([
    tsyringe_1.scoped(tsyringe_1.Lifecycle.ContainerScoped),
    __metadata("design:paramtypes", [AgentConfig_1.AgentConfig, IndyWallet_1.IndyWallet])
], IndyIssuerService);
exports.IndyIssuerService = IndyIssuerService;
