// @aries-framework/rest@0.8.0/build/controllers/proofs/ProofController.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProofController = void 0;
const core_1 = require("@aries-framework/core");
const JsonEncoder_1 = require("@aries-framework/core/build/utils/JsonEncoder");
const routing_controllers_1 = require("routing-controllers");
const typedi_1 = require("typedi");
const AcceptProofProposalRequest_1 = require("../../schemas/AcceptProofProposalRequest");
const PresentationProofRequest_1 = require("../../schemas/PresentationProofRequest");
const ProofPresentationRequest_1 = require("../../schemas/ProofPresentationRequest");
const ProofProposalRequest_1 = require("../../schemas/ProofProposalRequest");
const ProofRequestTemplate_1 = require("../../schemas/ProofRequestTemplate");
let ProofController = class ProofController {
    constructor(agent) {
        this.agent = agent;
    }
    /**
     * Retrieve proof record by proofRecordId
     */
    async getProofById(proofRecordId) {
        try {
            const proof = await this.agent.proofs.getById(proofRecordId);
            return proof.toJSON();
        }
        catch (error) {
            if (error instanceof core_1.RecordNotFoundError) {
                throw new routing_controllers_1.NotFoundError(`proof with proofRecordId "${proofRecordId}" not found.`);
            }
            throw new routing_controllers_1.InternalServerError(`something went wrong: ${error}`);
        }
    }
    /**
     * Retrieve all ProofRecords
     */
    async getAllProofs(threadId) {
        const proofs = await this.agent.proofs.getAll();
        if (threadId) {
            return proofs.flatMap((proof) => (proof.threadId === threadId ? proof.toJSON() : []));
        }
        return proofs.map((proof) => proof.toJSON());
    }
    /**
     * Initiate a new presentation exchange as prover by sending a presentation proposal message
     * to the connection with the specified connection id.
     */
    async proposeProof(proposal) {
        const { attributes, predicates, connectionId } = proposal, proposalOptions = __rest(proposal, ["attributes", "predicates", "connectionId"]);
        try {
            const presentationPreview = new core_1.PresentationPreview({
                attributes,
                predicates,
            });
            const proof = await this.agent.proofs.proposeProof(connectionId, presentationPreview, proposalOptions);
            return proof.toJSON();
        }
        catch (error) {
            if (error instanceof core_1.RecordNotFoundError) {
                throw new routing_controllers_1.NotFoundError(`connection with connectionId "${connectionId}" not found.`);
            }
            throw new routing_controllers_1.InternalServerError(`something went wrong: ${error}`);
        }
    }
    /**
     * Accept a presentation proposal as verifier (by sending a presentation request message) to the connection
     * associated with the proof record.
     */
    async acceptProposal(proofRecordId, proposal) {
        try {
            const proof = await this.agent.proofs.acceptProposal(proofRecordId, proposal);
            return proof.toJSON();
        }
        catch (error) {
            if (error instanceof core_1.RecordNotFoundError) {
                throw new routing_controllers_1.NotFoundError(`proof with proofRecordId "${proofRecordId}" not found.`);
            }
            throw new routing_controllers_1.InternalServerError(`something went wrong: ${error}`);
        }
    }
    /**
     * Creates a presentation request not bound to any proposal or existing connection
     */
    async requestProofOutOfBand(request) {
        const { proofRequest } = request, requestOptions = __rest(request, ["proofRequest"]);
        const proof = await this.agent.proofs.createOutOfBandRequest(proofRequest, requestOptions);
        return {
            message: `${this.agent.config.endpoints[0]}/?d_m=${JsonEncoder_1.JsonEncoder.toBase64URL(proof.requestMessage.toJSON({ useLegacyDidSovPrefix: this.agent.config.useLegacyDidSovPrefix }))}`,
            proofRecord: proof.proofRecord,
        };
    }
    /**
     * Creates a presentation request bound to existing connection
     */
    async requestProof(request) {
        const { connectionId, proofRequest } = request, requestOptions = __rest(request, ["connectionId", "proofRequest"]);
        try {
            const proof = await this.agent.proofs.requestProof(connectionId, proofRequest, requestOptions);
            return proof.toJSON();
        }
        catch (error) {
            if (error instanceof core_1.RecordNotFoundError) {
                throw new routing_controllers_1.NotFoundError(`connection with connectionId "${connectionId}" not found.`);
            }
            throw new routing_controllers_1.InternalServerError(`something went wrong: ${error}`);
        }
    }
    /**
     * Accept a presentation request as prover (by sending a presentation message) to the connection
     * associated with the proof record.
     */
    async acceptRequest(proofRecordId, request) {
        try {
            const { filterByPresentationPreview, comment } = request;
            const retrievedCredentials = await this.agent.proofs.getRequestedCredentialsForProofRequest(proofRecordId, {
                filterByPresentationPreview: filterByPresentationPreview,
            });
            const requestedCredentials = this.agent.proofs.autoSelectCredentialsForProofRequest(retrievedCredentials);
            const proof = await this.agent.proofs.acceptRequest(proofRecordId, requestedCredentials, {
                comment,
            });
            return proof.toJSON();
        }
        catch (error) {
            if (error instanceof core_1.RecordNotFoundError) {
                throw new routing_controllers_1.NotFoundError(`Proof with proofRecordId "${proofRecordId}" not found.`);
            }
            throw new routing_controllers_1.InternalServerError(`Something went wrong: ${error}`);
        }
    }
    /**
     * Accept a presentation as prover (by sending a presentation acknowledgement message) to the connection
     * associated with the proof record.
     */
    async acceptPresentation(proofRecordId) {
        try {
            const proof = await this.agent.proofs.acceptPresentation(proofRecordId);
            return proof.toJSON();
        }
        catch (error) {
            if (error instanceof core_1.RecordNotFoundError) {
                throw new routing_controllers_1.NotFoundError(`Proof with proofRecordId "${proofRecordId}" not found.`);
            }
            throw new routing_controllers_1.InternalServerError(`Something went wrong: ${error}`);
        }
    }
    /**
     * Deletes a proofRecord in the proof repository.
     */
    async deleteProof(proofRecordId) {
        try {
            await this.agent.proofs.deleteById(proofRecordId);
        }
        catch (error) {
            if (error instanceof core_1.RecordNotFoundError) {
                throw new routing_controllers_1.NotFoundError(`Proof with proofRecordId "${proofRecordId}" not found.`);
            }
            throw new routing_controllers_1.InternalServerError(`Something went wrong: ${error}`);
        }
    }
};
__decorate([
    (0, typedi_1.Inject)(),
    __metadata("design:type", core_1.Agent)
], ProofController.prototype, "agent", void 0);
__decorate([
    (0, routing_controllers_1.Get)('/:proofRecordId'),
    __param(0, (0, routing_controllers_1.Param)('proofRecordId')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", Promise)
], ProofController.prototype, "getProofById", null);
__decorate([
    (0, routing_controllers_1.Get)('/'),
    __param(0, (0, routing_controllers_1.QueryParam)('threadId')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", Promise)
], ProofController.prototype, "getAllProofs", null);
__decorate([
    (0, routing_controllers_1.Post)('/propose-proof'),
    __param(0, (0, routing_controllers_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [ProofProposalRequest_1.ProofProposalRequest]),
    __metadata("design:returntype", Promise)
], ProofController.prototype, "proposeProof", null);
__decorate([
    (0, routing_controllers_1.Post)('/:proofRecordId/accept-proposal'),
    __param(0, (0, routing_controllers_1.Param)('proofRecordId')),
    __param(1, (0, routing_controllers_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, AcceptProofProposalRequest_1.AcceptProofProposalRequest]),
    __metadata("design:returntype", Promise)
], ProofController.prototype, "acceptProposal", null);
__decorate([
    (0, routing_controllers_1.Post)('/request-outofband-proof'),
    __param(0, (0, routing_controllers_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [ProofRequestTemplate_1.ProofRequestTemplate]),
    __metadata("design:returntype", Promise)
], ProofController.prototype, "requestProofOutOfBand", null);
__decorate([
    (0, routing_controllers_1.Post)('/request-proof'),
    __param(0, (0, routing_controllers_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [ProofPresentationRequest_1.ProofPresentationRequest]),
    __metadata("design:returntype", Promise)
], ProofController.prototype, "requestProof", null);
__decorate([
    (0, routing_controllers_1.Post)('/:proofRecordId/accept-request'),
    __param(0, (0, routing_controllers_1.Param)('proofRecordId')),
    __param(1, (0, routing_controllers_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, PresentationProofRequest_1.PresentationProofRequest]),
    __metadata("design:returntype", Promise)
], ProofController.prototype, "acceptRequest", null);
__decorate([
    (0, routing_controllers_1.Post)('/:proofRecordId/accept-presentation'),
    __param(0, (0, routing_controllers_1.Param)('proofRecordId')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", Promise)
], ProofController.prototype, "acceptPresentation", null);
__decorate([
    (0, routing_controllers_1.Delete)('/:proofRecordId'),
    (0, routing_controllers_1.OnUndefined)(204),
    __param(0, (0, routing_controllers_1.Param)('proofRecordId')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", Promise)
], ProofController.prototype, "deleteProof", null);
ProofController = __decorate([
    (0, routing_controllers_1.JsonController)('/proofs'),
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [core_1.Agent])
], ProofController);
exports.ProofController = ProofController;
