// @aries-framework/core@0.1.0/build/modules/connections/models/did/service/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Service = exports.DidCommService = exports.IndyAgentService = exports.ServiceTransformer = exports.serviceTypes = void 0;
const class_transformer_1 = require("class-transformer");
const DidCommService_1 = require("./DidCommService");
Object.defineProperty(exports, "DidCommService", { enumerable: true, get: function () { return DidCommService_1.DidCommService; } });
const IndyAgentService_1 = require("./IndyAgentService");
Object.defineProperty(exports, "IndyAgentService", { enumerable: true, get: function () { return IndyAgentService_1.IndyAgentService; } });
const Service_1 = require("./Service");
Object.defineProperty(exports, "Service", { enumerable: true, get: function () { return Service_1.Service; } });
exports.serviceTypes = {
    [IndyAgentService_1.IndyAgentService.type]: IndyAgentService_1.IndyAgentService,
    [DidCommService_1.DidCommService.type]: DidCommService_1.DidCommService,
};
/**
 * Decorator that transforms service json to corresponding class instances. See {@link serviceTypes}
 *
 * @example
 * class Example {
 *   ServiceTransformer()
 *   private service: Service
 * }
 */
function ServiceTransformer() {
    return class_transformer_1.Transform(({ value }) => {
        return value.map((serviceJson) => {
            var _a;
            const serviceClass = ((_a = exports.serviceTypes[serviceJson.type]) !== null && _a !== void 0 ? _a : Service_1.Service);
            const service = class_transformer_1.plainToInstance(serviceClass, serviceJson);
            return service;
        });
    }, {
        toClassOnly: true,
    });
}
exports.ServiceTransformer = ServiceTransformer;
