// @aries-framework/node@0.1.0/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WsInboundTransport = exports.HttpInboundTransport = exports.agentDependencies = void 0;
const events_1 = require("events");
const indy = __importStar(require("indy-sdk"));
const node_fetch_1 = __importDefault(require("node-fetch"));
const ws_1 = __importDefault(require("ws"));
const NodeFileSystem_1 = require("./NodeFileSystem");
const HttpInboundTransport_1 = require("./transport/HttpInboundTransport");
Object.defineProperty(exports, "HttpInboundTransport", { enumerable: true, get: function () { return HttpInboundTransport_1.HttpInboundTransport; } });
const WsInboundTransport_1 = require("./transport/WsInboundTransport");
Object.defineProperty(exports, "WsInboundTransport", { enumerable: true, get: function () { return WsInboundTransport_1.WsInboundTransport; } });
const agentDependencies = {
    FileSystem: NodeFileSystem_1.NodeFileSystem,
    fetch: node_fetch_1.default,
    EventEmitterClass: events_1.EventEmitter,
    WebSocketClass: ws_1.default,
    indy,
};
exports.agentDependencies = agentDependencies;
