// @aries-framework/node@0.1.0/build/NodeFileSystem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeFileSystem = void 0;
const fs_1 = require("fs");
const os_1 = require("os");
const path_1 = require("path");
const { access, readFile, writeFile } = fs_1.promises;
class NodeFileSystem {
    /**
     * Create new NodeFileSystem class instance.
     *
     * @param basePath The base path to use for reading and writing files. process.cwd() if not specified
     */
    constructor(basePath) {
        this.basePath = basePath !== null && basePath !== void 0 ? basePath : os_1.tmpdir();
    }
    async exists(path) {
        try {
            await access(path);
            return true;
        }
        catch (_a) {
            return false;
        }
    }
    async write(path, data) {
        // Make sure parent directories exist
        await fs_1.promises.mkdir(path_1.dirname(path), { recursive: true });
        return writeFile(path, data, { encoding: 'utf-8' });
    }
    async read(path) {
        return readFile(path, { encoding: 'utf-8' });
    }
}
exports.NodeFileSystem = NodeFileSystem;
