// @aries-framework/core@0.1.0/build/modules/proofs/messages/RequestPresentationMessage.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestPresentationMessage = exports.INDY_PROOF_REQUEST_ATTACHMENT_ID = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const AgentMessage_1 = require("../../../agent/AgentMessage");
const Attachment_1 = require("../../../decorators/attachment/Attachment");
const JsonTransformer_1 = require("../../../utils/JsonTransformer");
const models_1 = require("../models");
exports.INDY_PROOF_REQUEST_ATTACHMENT_ID = 'libindy-request-presentation-0';
/**
 * Request Presentation Message part of Present Proof Protocol used to initiate request from verifier to prover.
 *
 * @see https://github.com/hyperledger/aries-rfcs/blob/master/features/0037-present-proof/README.md#request-presentation
 */
class RequestPresentationMessage extends AgentMessage_1.AgentMessage {
    constructor(options) {
        var _a;
        super();
        this.type = RequestPresentationMessage.type;
        if (options) {
            this.id = (_a = options.id) !== null && _a !== void 0 ? _a : this.generateId();
            this.comment = options.comment;
            this.requestPresentationAttachments = options.requestPresentationAttachments;
        }
    }
    get indyProofRequest() {
        var _a, _b;
        const attachment = this.requestPresentationAttachments.find((attachment) => attachment.id === exports.INDY_PROOF_REQUEST_ATTACHMENT_ID);
        // Extract proof request from attachment
        const proofRequestJson = (_b = (_a = attachment === null || attachment === void 0 ? void 0 : attachment.data) === null || _a === void 0 ? void 0 : _a.getDataAsJson()) !== null && _b !== void 0 ? _b : null;
        const proofRequest = JsonTransformer_1.JsonTransformer.fromJSON(proofRequestJson, models_1.ProofRequest);
        return proofRequest;
    }
}
RequestPresentationMessage.type = 'https://didcomm.org/present-proof/1.0/request-presentation';
__decorate([
    class_validator_1.Equals(RequestPresentationMessage.type),
    __metadata("design:type", Object)
], RequestPresentationMessage.prototype, "type", void 0);
__decorate([
    class_validator_1.IsOptional(),
    class_validator_1.IsString(),
    __metadata("design:type", String)
], RequestPresentationMessage.prototype, "comment", void 0);
__decorate([
    class_transformer_1.Expose({ name: 'request_presentations~attach' }),
    class_transformer_1.Type(() => Attachment_1.Attachment),
    class_validator_1.IsArray(),
    class_validator_1.ValidateNested({
        each: true,
    }),
    class_validator_1.IsInstance(Attachment_1.Attachment, { each: true }),
    __metadata("design:type", Array)
], RequestPresentationMessage.prototype, "requestPresentationAttachments", void 0);
exports.RequestPresentationMessage = RequestPresentationMessage;
