// @aries-framework/core@0.1.0/build/modules/credentials/handlers/IssueCredentialHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IssueCredentialHandler = void 0;
const helpers_1 = require("../../../agent/helpers");
const messages_1 = require("../messages");
class IssueCredentialHandler {
    constructor(credentialService, agentConfig, credentialResponseCoordinator) {
        this.supportedMessages = [messages_1.IssueCredentialMessage];
        this.credentialService = credentialService;
        this.agentConfig = agentConfig;
        this.credentialResponseCoordinator = credentialResponseCoordinator;
    }
    async handle(messageContext) {
        const credentialRecord = await this.credentialService.processCredential(messageContext);
        if (this.credentialResponseCoordinator.shouldAutoRespondToIssue(credentialRecord)) {
            return await this.createAck(credentialRecord, messageContext);
        }
    }
    async createAck(record, messageContext) {
        var _a, _b;
        this.agentConfig.logger.info(`Automatically sending acknowledgement with autoAccept on ${this.agentConfig.autoAcceptCredentials}`);
        const { message, credentialRecord } = await this.credentialService.createAck(record);
        if (messageContext.connection) {
            return helpers_1.createOutboundMessage(messageContext.connection, message);
        }
        else if (((_a = credentialRecord.credentialMessage) === null || _a === void 0 ? void 0 : _a.service) && ((_b = credentialRecord.requestMessage) === null || _b === void 0 ? void 0 : _b.service)) {
            const recipientService = credentialRecord.credentialMessage.service;
            const ourService = credentialRecord.requestMessage.service;
            return helpers_1.createOutboundServiceMessage({
                payload: message,
                service: recipientService.toDidCommService(),
                senderKey: ourService.recipientKeys[0],
            });
        }
        this.agentConfig.logger.error(`Could not automatically create credential ack`);
    }
}
exports.IssueCredentialHandler = IssueCredentialHandler;
