// @aries-framework/core@0.1.0/build/transport/HttpOutboundTransport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpOutboundTransport = void 0;
const abort_controller_1 = require("abort-controller");
const AgentConfig_1 = require("../agent/AgentConfig");
const AriesFrameworkError_1 = require("../error/AriesFrameworkError");
class HttpOutboundTransport {
    constructor() {
        this.supportedSchemes = ['http', 'https'];
    }
    async start(agent) {
        this.agent = agent;
        this.agentConfig = agent.injectionContainer.resolve(AgentConfig_1.AgentConfig);
        this.logger = this.agentConfig.logger;
        this.fetch = this.agentConfig.agentDependencies.fetch;
        this.logger.debug('Starting HTTP outbound transport');
    }
    async stop() {
        this.logger.debug('Stopping HTTP outbound transport');
        // Nothing required to stop HTTP
    }
    async sendMessage(outboundPackage) {
        const { payload, endpoint } = outboundPackage;
        if (!endpoint) {
            throw new AriesFrameworkError_1.AriesFrameworkError(`Missing endpoint. I don't know how and where to send the message.`);
        }
        this.logger.debug(`Sending outbound message to endpoint '${outboundPackage.endpoint}'`, {
            payload: outboundPackage.payload,
        });
        try {
            const abortController = new abort_controller_1.AbortController();
            const id = setTimeout(() => abortController.abort(), 15000);
            let response;
            let responseMessage;
            try {
                response = await this.fetch(endpoint, {
                    method: 'POST',
                    body: JSON.stringify(payload),
                    headers: { 'Content-Type': this.agentConfig.didCommMimeType },
                    signal: abortController.signal,
                });
                clearTimeout(id);
                responseMessage = await response.text();
            }
            catch (error) {
                // Request is aborted after 15 seconds, but that doesn't necessarily mean the request
                // went wrong. ACA-Py keeps the socket alive until it has a response message. So we assume
                // that if the error was aborted and we had return routing enabled, we should ignore the error.
                if (error.name == 'AbortError' && outboundPackage.responseRequested) {
                    this.logger.debug('Request was aborted due to timeout. Not throwing error due to return routing on sent message');
                }
                else {
                    throw error;
                }
            }
            // TODO: do we just want to ignore messages that were returned if we didn't request it?
            // TODO: check response header type (and also update inbound transports to use the correct headers types)
            if (response && responseMessage) {
                this.logger.debug(`Response received`, { responseMessage, status: response.status });
                try {
                    const encryptedMessage = JSON.parse(responseMessage);
                    this.agent.receiveMessage(encryptedMessage);
                }
                catch (error) {
                    this.logger.debug('Unable to parse response message');
                }
            }
            else {
                this.logger.debug(`No response received.`);
            }
        }
        catch (error) {
            this.logger.error(`Error sending message to ${endpoint}: ${error.message}`, {
                error,
                message: error.message,
                body: payload,
                didCommMimeType: this.agentConfig.didCommMimeType,
            });
            throw new AriesFrameworkError_1.AriesFrameworkError(`Error sending message to ${endpoint}: ${error.message}`, { cause: error });
        }
    }
}
exports.HttpOutboundTransport = HttpOutboundTransport;
