// @aries-framework/core@0.1.0/build/modules/discover-features/services/DiscoverFeaturesService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiscoverFeaturesService = void 0;
const tsyringe_1 = require("tsyringe");
const Dispatcher_1 = require("../../../agent/Dispatcher");
const messages_1 = require("../messages");
let DiscoverFeaturesService = class DiscoverFeaturesService {
    constructor(dispatcher) {
        this.dispatcher = dispatcher;
    }
    async createQuery(options) {
        const queryMessage = new messages_1.QueryMessage(options);
        return queryMessage;
    }
    async createDisclose(queryMessage) {
        const { query } = queryMessage;
        const messageFamilies = this.dispatcher.supportedProtocols;
        let protocols = [];
        if (query === '*') {
            protocols = messageFamilies;
        }
        else if (query.endsWith('*')) {
            const match = query.slice(0, -1);
            protocols = messageFamilies.filter((m) => m.startsWith(match));
        }
        else if (messageFamilies.includes(query)) {
            protocols = [query];
        }
        const discloseMessage = new messages_1.DiscloseMessage({
            threadId: queryMessage.threadId,
            protocols: protocols.map((protocolId) => ({ protocolId })),
        });
        return discloseMessage;
    }
};
DiscoverFeaturesService = __decorate([
    tsyringe_1.scoped(tsyringe_1.Lifecycle.ContainerScoped),
    __metadata("design:paramtypes", [Dispatcher_1.Dispatcher])
], DiscoverFeaturesService);
exports.DiscoverFeaturesService = DiscoverFeaturesService;
