// @aries-framework/core@0.1.0/build/modules/discover-features/DiscoverFeaturesModule.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiscoverFeaturesModule = void 0;
const tsyringe_1 = require("tsyringe");
const Dispatcher_1 = require("../../agent/Dispatcher");
const MessageSender_1 = require("../../agent/MessageSender");
const helpers_1 = require("../../agent/helpers");
const services_1 = require("../connections/services");
const handlers_1 = require("./handlers");
const services_2 = require("./services");
let DiscoverFeaturesModule = class DiscoverFeaturesModule {
    constructor(dispatcher, connectionService, messageSender, discoverFeaturesService) {
        this.connectionService = connectionService;
        this.messageSender = messageSender;
        this.discoverFeaturesService = discoverFeaturesService;
        this.registerHandlers(dispatcher);
    }
    async queryFeatures(connectionId, options) {
        const connection = await this.connectionService.getById(connectionId);
        const queryMessage = await this.discoverFeaturesService.createQuery(options);
        const outbound = helpers_1.createOutboundMessage(connection, queryMessage);
        await this.messageSender.sendMessage(outbound);
    }
    registerHandlers(dispatcher) {
        dispatcher.registerHandler(new handlers_1.DiscloseMessageHandler());
        dispatcher.registerHandler(new handlers_1.QueryMessageHandler(this.discoverFeaturesService));
    }
};
DiscoverFeaturesModule = __decorate([
    tsyringe_1.scoped(tsyringe_1.Lifecycle.ContainerScoped),
    __metadata("design:paramtypes", [Dispatcher_1.Dispatcher,
        services_1.ConnectionService,
        MessageSender_1.MessageSender,
        services_2.DiscoverFeaturesService])
], DiscoverFeaturesModule);
exports.DiscoverFeaturesModule = DiscoverFeaturesModule;
