// @aries-framework/core@0.1.0/build/modules/ledger/services/IndyLedgerService.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndyLedgerService = void 0;
const tsyringe_1 = require("tsyringe");
const AgentConfig_1 = require("../../../agent/AgentConfig");
const IndySdkError_1 = require("../../../error/IndySdkError");
const did_1 = require("../../../utils/did");
const indyError_1 = require("../../../utils/indyError");
const IndyWallet_1 = require("../../../wallet/IndyWallet");
const indy_1 = require("../../indy");
const IndyPoolService_1 = require("./IndyPoolService");
let IndyLedgerService = class IndyLedgerService {
    constructor(wallet, agentConfig, indyIssuer, indyPoolService) {
        this.wallet = wallet;
        this.indy = agentConfig.agentDependencies.indy;
        this.logger = agentConfig.logger;
        this.indyIssuer = indyIssuer;
        this.indyPoolService = indyPoolService;
    }
    async registerPublicDid(submitterDid, targetDid, verkey, alias, role) {
        const pool = this.indyPoolService.ledgerWritePool;
        try {
            this.logger.debug(`Register public did '${targetDid}' on ledger '${pool.id}'`);
            const request = await this.indy.buildNymRequest(submitterDid, targetDid, verkey, alias, role || null);
            const response = await this.submitWriteRequest(pool, request, submitterDid);
            this.logger.debug(`Registered public did '${targetDid}' on ledger '${pool.id}'`, {
                response,
            });
            return targetDid;
        }
        catch (error) {
            this.logger.error(`Error registering public did '${targetDid}' on ledger '${pool.id}'`, {
                error,
                submitterDid,
                targetDid,
                verkey,
                alias,
                role,
                pool,
            });
            throw error;
        }
    }
    async getPublicDid(did) {
        // Getting the pool for a did also retrieves the DID. We can just use that
        const { did: didResponse } = await this.indyPoolService.getPoolForDid(did);
        return didResponse;
    }
    async registerSchema(did, schemaTemplate) {
        const pool = this.indyPoolService.ledgerWritePool;
        try {
            this.logger.debug(`Register schema on ledger '${pool.id}' with did '${did}'`, schemaTemplate);
            const { name, attributes, version } = schemaTemplate;
            const schema = await this.indyIssuer.createSchema({ originDid: did, name, version, attributes });
            const request = await this.indy.buildSchemaRequest(did, schema);
            const response = await this.submitWriteRequest(pool, request, did);
            this.logger.debug(`Registered schema '${schema.id}' on ledger '${pool.id}'`, {
                response,
                schema,
            });
            schema.seqNo = response.result.txnMetadata.seqNo;
            return schema;
        }
        catch (error) {
            this.logger.error(`Error registering schema for did '${did}' on ledger '${pool.id}'`, {
                error,
                did,
                schemaTemplate,
            });
            throw indyError_1.isIndyError(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async getSchema(schemaId) {
        const did = did_1.didFromSchemaId(schemaId);
        const { pool } = await this.indyPoolService.getPoolForDid(did);
        try {
            this.logger.debug(`Get schema '${schemaId}' from ledger '${pool.id}'`);
            const request = await this.indy.buildGetSchemaRequest(null, schemaId);
            this.logger.debug(`Submitting get schema request for schema '${schemaId}' to ledger '${pool.id}'`);
            const response = await this.submitReadRequest(pool, request);
            const [, schema] = await this.indy.parseGetSchemaResponse(response);
            this.logger.debug(`Got schema '${schemaId}' from ledger '${pool.id}'`, {
                response,
                schema,
            });
            return schema;
        }
        catch (error) {
            this.logger.error(`Error retrieving schema '${schemaId}' from ledger '${pool.id}'`, {
                error,
                schemaId,
            });
            throw indyError_1.isIndyError(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async registerCredentialDefinition(did, credentialDefinitionTemplate) {
        const pool = this.indyPoolService.ledgerWritePool;
        try {
            this.logger.debug(`Register credential definition on ledger '${pool.id}' with did '${did}'`, credentialDefinitionTemplate);
            const { schema, tag, signatureType, supportRevocation } = credentialDefinitionTemplate;
            const credentialDefinition = await this.indyIssuer.createCredentialDefinition({
                issuerDid: did,
                schema,
                tag,
                signatureType,
                supportRevocation,
            });
            const request = await this.indy.buildCredDefRequest(did, credentialDefinition);
            const response = await this.submitWriteRequest(pool, request, did);
            this.logger.debug(`Registered credential definition '${credentialDefinition.id}' on ledger '${pool.id}'`, {
                response,
                credentialDefinition: credentialDefinition,
            });
            return credentialDefinition;
        }
        catch (error) {
            this.logger.error(`Error registering credential definition for schema '${credentialDefinitionTemplate.schema.id}' on ledger '${pool.id}'`, {
                error,
                did,
                credentialDefinitionTemplate,
            });
            throw indyError_1.isIndyError(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async getCredentialDefinition(credentialDefinitionId) {
        const did = did_1.didFromCredentialDefinitionId(credentialDefinitionId);
        const { pool } = await this.indyPoolService.getPoolForDid(did);
        this.logger.debug(`Using ledger '${pool.id}' to retrieve credential definition '${credentialDefinitionId}'`);
        try {
            this.logger.debug(`Get credential definition '${credentialDefinitionId}' from ledger '${pool.id}'`);
            const request = await this.indy.buildGetCredDefRequest(null, credentialDefinitionId);
            this.logger.debug(`Submitting get credential definition request for credential definition '${credentialDefinitionId}' to ledger '${pool.id}'`);
            const response = await this.submitReadRequest(pool, request);
            const [, credentialDefinition] = await this.indy.parseGetCredDefResponse(response);
            this.logger.debug(`Got credential definition '${credentialDefinitionId}' from ledger '${pool.id}'`, {
                response,
                credentialDefinition,
            });
            return credentialDefinition;
        }
        catch (error) {
            this.logger.error(`Error retrieving credential definition '${credentialDefinitionId}' from ledger '${pool.id}'`, {
                error,
                credentialDefinitionId,
                pool: pool.id,
            });
            throw indyError_1.isIndyError(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async submitWriteRequest(pool, request, signDid) {
        try {
            const requestWithTaa = await this.appendTaa(pool, request);
            const signedRequestWithTaa = await this.signRequest(signDid, requestWithTaa);
            const response = await pool.submitWriteRequest(signedRequestWithTaa);
            return response;
        }
        catch (error) {
            throw indyError_1.isIndyError(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async submitReadRequest(pool, request) {
        try {
            const response = await pool.submitReadRequest(request);
            return response;
        }
        catch (error) {
            throw indyError_1.isIndyError(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async signRequest(did, request) {
        try {
            return this.indy.signRequest(this.wallet.handle, did, request);
        }
        catch (error) {
            throw indyError_1.isIndyError(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async appendTaa(pool, request) {
        try {
            const authorAgreement = await this.getTransactionAuthorAgreement(pool);
            // If ledger does not have TAA, we can just send request
            if (authorAgreement == null) {
                return request;
            }
            const requestWithTaa = await this.indy.appendTxnAuthorAgreementAcceptanceToRequest(request, authorAgreement.text, authorAgreement.version, authorAgreement.digest, this.getFirstAcceptanceMechanism(authorAgreement), 
            // Current time since epoch
            // We can't use ratification_ts, as it must be greater than 1499906902
            Math.floor(new Date().getTime() / 1000));
            return requestWithTaa;
        }
        catch (error) {
            throw indyError_1.isIndyError(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    async getTransactionAuthorAgreement(pool) {
        try {
            // TODO Replace this condition with memoization
            if (pool.authorAgreement !== undefined) {
                return pool.authorAgreement;
            }
            const taaRequest = await this.indy.buildGetTxnAuthorAgreementRequest(null);
            const taaResponse = await this.submitReadRequest(pool, taaRequest);
            const acceptanceMechanismRequest = await this.indy.buildGetAcceptanceMechanismsRequest(null);
            const acceptanceMechanismResponse = await this.submitReadRequest(pool, acceptanceMechanismRequest);
            // TAA can be null
            if (taaResponse.result.data == null) {
                pool.authorAgreement = null;
                return null;
            }
            // If TAA is not null, we can be sure AcceptanceMechanisms is also not null
            const authorAgreement = taaResponse.result.data;
            const acceptanceMechanisms = acceptanceMechanismResponse.result.data;
            pool.authorAgreement = Object.assign(Object.assign({}, authorAgreement), { acceptanceMechanisms });
            return pool.authorAgreement;
        }
        catch (error) {
            throw indyError_1.isIndyError(error) ? new IndySdkError_1.IndySdkError(error) : error;
        }
    }
    getFirstAcceptanceMechanism(authorAgreement) {
        const [firstMechanism] = Object.keys(authorAgreement.acceptanceMechanisms.aml);
        return firstMechanism;
    }
};
IndyLedgerService = __decorate([
    tsyringe_1.scoped(tsyringe_1.Lifecycle.ContainerScoped),
    __metadata("design:paramtypes", [IndyWallet_1.IndyWallet,
        AgentConfig_1.AgentConfig,
        indy_1.IndyIssuerService,
        IndyPoolService_1.IndyPoolService])
], IndyLedgerService);
exports.IndyLedgerService = IndyLedgerService;
