// @aries-framework/core@0.1.0/build/modules/routing/handlers/BatchHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BatchHandler = void 0;
const Events_1 = require("../../../agent/Events");
const error_1 = require("../../../error");
const messages_1 = require("../messages");
class BatchHandler {
    constructor(eventEmitter) {
        this.supportedMessages = [messages_1.BatchMessage];
        this.eventEmitter = eventEmitter;
    }
    async handle(messageContext) {
        const { message, connection } = messageContext;
        if (!connection) {
            throw new error_1.AriesFrameworkError(`No connection associated with incoming message with id ${message.id}`);
        }
        const forwardedMessages = message.messages;
        forwardedMessages.forEach((message) => {
            this.eventEmitter.emit({
                type: Events_1.AgentEventTypes.AgentMessageReceived,
                payload: {
                    message: message.message,
                },
            });
        });
    }
}
exports.BatchHandler = BatchHandler;
