// @aries-framework/rest@0.8.0/build/controllers/credentials/SchemaController.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SchemaController = void 0;
const core_1 = require("@aries-framework/core");
const LedgerError_1 = require("@aries-framework/core/build/modules/ledger/error/LedgerError");
const LedgerNotFoundError_1 = require("@aries-framework/core/build/modules/ledger/error/LedgerNotFoundError");
const indyError_1 = require("@aries-framework/core/build/utils/indyError");
const routing_controllers_1 = require("routing-controllers");
const typedi_1 = require("typedi");
const SchemaRequest_1 = require("../../schemas/SchemaRequest");
let SchemaController = class SchemaController {
    constructor(agent) {
        this.agent = agent;
    }
    /**
     * Retrieve schema by schemaId
     */
    async getSchemaById(schemaId) {
        try {
            return await this.agent.ledger.getSchema(schemaId);
        }
        catch (error) {
            if (error instanceof LedgerNotFoundError_1.LedgerNotFoundError) {
                throw new routing_controllers_1.NotFoundError(`schema definition with schemaId "${schemaId}" not found.`);
            }
            else if (error instanceof LedgerError_1.LedgerError && error.cause instanceof core_1.IndySdkError) {
                if ((0, indyError_1.isIndyError)(error.cause.cause, 'LedgerInvalidTransaction')) {
                    throw new routing_controllers_1.ForbiddenError(`schema definition with schemaId "${schemaId}" can not be returned.`);
                }
                if ((0, indyError_1.isIndyError)(error.cause.cause, 'CommonInvalidStructure')) {
                    throw new routing_controllers_1.BadRequestError(`schemaId "${schemaId}" has invalid structure.`);
                }
            }
            throw new routing_controllers_1.InternalServerError(`something went wrong: ${error}`);
        }
    }
    /**
     * Creates a new schema and registers schema on ledger
     * Returns created schema
     */
    async createSchema(schema) {
        try {
            return await this.agent.ledger.registerSchema({
                name: schema.name,
                version: schema.version,
                attributes: schema.attributes,
            });
        }
        catch (error) {
            if (error instanceof core_1.AriesFrameworkError) {
                if (error.message.includes('UnauthorizedClientRequest')) {
                    throw new routing_controllers_1.ForbiddenError(`this action is not allowed.`);
                }
            }
            throw new routing_controllers_1.InternalServerError(`something went wrong: ${error}`);
        }
    }
};
__decorate([
    (0, typedi_1.Inject)(),
    __metadata("design:type", core_1.Agent)
], SchemaController.prototype, "agent", void 0);
__decorate([
    (0, routing_controllers_1.Get)('/:schemaId'),
    __param(0, (0, routing_controllers_1.Param)('schemaId')),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String]),
    __metadata("design:returntype", Promise)
], SchemaController.prototype, "getSchemaById", null);
__decorate([
    (0, routing_controllers_1.Post)('/'),
    __param(0, (0, routing_controllers_1.Body)()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [SchemaRequest_1.SchemaTemplate]),
    __metadata("design:returntype", Promise)
], SchemaController.prototype, "createSchema", null);
SchemaController = __decorate([
    (0, routing_controllers_1.JsonController)('/schemas'),
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [core_1.Agent])
], SchemaController);
exports.SchemaController = SchemaController;
