// @aries-framework/core@0.1.0/build/modules/discover-features/messages/DiscloseMessage.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiscloseMessage = exports.DiscloseProtocol = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const AgentMessage_1 = require("../../../agent/AgentMessage");
class DiscloseProtocol {
    constructor(options) {
        if (options) {
            this.protocolId = options.protocolId;
            this.roles = options.roles;
        }
    }
}
__decorate([
    class_transformer_1.Expose({ name: 'pid' }),
    class_validator_1.IsString(),
    __metadata("design:type", String)
], DiscloseProtocol.prototype, "protocolId", void 0);
__decorate([
    class_validator_1.IsString({ each: true }),
    class_validator_1.IsOptional(),
    __metadata("design:type", Array)
], DiscloseProtocol.prototype, "roles", void 0);
exports.DiscloseProtocol = DiscloseProtocol;
class DiscloseMessage extends AgentMessage_1.AgentMessage {
    constructor(options) {
        var _a;
        super();
        this.type = DiscloseMessage.type;
        if (options) {
            this.id = (_a = options.id) !== null && _a !== void 0 ? _a : this.generateId();
            this.protocols = options.protocols.map((p) => new DiscloseProtocol(p));
            this.setThread({
                threadId: options.threadId,
            });
        }
    }
}
DiscloseMessage.type = 'https://didcomm.org/discover-features/1.0/disclose';
__decorate([
    class_validator_1.Equals(DiscloseMessage.type),
    __metadata("design:type", Object)
], DiscloseMessage.prototype, "type", void 0);
__decorate([
    class_validator_1.IsInstance(DiscloseProtocol, { each: true }),
    class_transformer_1.Type(() => DiscloseProtocol),
    __metadata("design:type", Array)
], DiscloseMessage.prototype, "protocols", void 0);
exports.DiscloseMessage = DiscloseMessage;
