// @aries-framework/core@0.1.0/build/modules/connections/messages/ConnectionInvitationMessage.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectionInvitationMessage = void 0;
const class_transformer_1 = require("class-transformer");
const class_validator_1 = require("class-validator");
const query_string_1 = require("query-string");
const AgentMessage_1 = require("../../../agent/AgentMessage");
const error_1 = require("../../../error");
const JsonEncoder_1 = require("../../../utils/JsonEncoder");
const JsonTransformer_1 = require("../../../utils/JsonTransformer");
const MessageValidator_1 = require("../../../utils/MessageValidator");
const messageType_1 = require("../../../utils/messageType");
/**
 * Message to invite another agent to create a connection
 *
 * @see https://github.com/hyperledger/aries-rfcs/blob/master/features/0160-connection-protocol/README.md#0-invitation-to-connect
 */
class ConnectionInvitationMessage extends AgentMessage_1.AgentMessage {
    /**
     * Create new ConnectionInvitationMessage instance.
     * @param options
     */
    constructor(options) {
        super();
        this.type = ConnectionInvitationMessage.type;
        if (options) {
            this.id = options.id || this.generateId();
            this.label = options.label;
            this.imageUrl = options.imageUrl;
            if (isDidInvitation(options)) {
                this.did = options.did;
            }
            else {
                this.recipientKeys = options.recipientKeys;
                this.serviceEndpoint = options.serviceEndpoint;
                this.routingKeys = options.routingKeys;
            }
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            if (options.did && (options.recipientKeys || options.routingKeys || options.serviceEndpoint)) {
                throw new error_1.AriesFrameworkError('either the did or the recipientKeys/serviceEndpoint/routingKeys must be set, but not both');
            }
        }
    }
    /**
     * Create an invitation url from this instance
     *
     * @param domain domain name to use for invitation url
     * @returns invitation url with base64 encoded invitation
     */
    toUrl({ domain, useLegacyDidSovPrefix = false }) {
        const invitationJson = this.toJSON({ useLegacyDidSovPrefix });
        const encodedInvitation = JsonEncoder_1.JsonEncoder.toBase64URL(invitationJson);
        const invitationUrl = `${domain}?c_i=${encodedInvitation}`;
        return invitationUrl;
    }
    /**
     * Create a `ConnectionInvitationMessage` instance from the `c_i` or `d_m` parameter of an URL
     *
     * @param invitationUrl invitation url containing c_i or d_m parameter
     *
     * @throws Error when url can not be decoded to JSON, or decoded message is not a valid `ConnectionInvitationMessage`
     * @throws Error when the url does not contain c_i or d_m as parameter
     */
    static async fromUrl(invitationUrl) {
        var _a;
        const parsedUrl = query_string_1.parseUrl(invitationUrl).query;
        const encodedInvitation = (_a = parsedUrl['c_i']) !== null && _a !== void 0 ? _a : parsedUrl['d_m'];
        if (typeof encodedInvitation === 'string') {
            const invitationJson = JsonEncoder_1.JsonEncoder.fromBase64(encodedInvitation);
            const invitation = JsonTransformer_1.JsonTransformer.fromJSON(invitationJson, ConnectionInvitationMessage);
            await MessageValidator_1.MessageValidator.validate(invitation);
            return invitation;
        }
        else {
            throw new error_1.AriesFrameworkError('InvitationUrl is invalid. It needs to contain one, and only one, of the following parameters; `c_i` or `d_m`');
        }
    }
}
ConnectionInvitationMessage.type = 'https://didcomm.org/connections/1.0/invitation';
__decorate([
    class_validator_1.Equals(ConnectionInvitationMessage.type),
    class_transformer_1.Transform(({ value }) => messageType_1.replaceLegacyDidSovPrefix(value), {
        toClassOnly: true,
    }),
    __metadata("design:type", Object)
], ConnectionInvitationMessage.prototype, "type", void 0);
__decorate([
    class_validator_1.IsString(),
    __metadata("design:type", String)
], ConnectionInvitationMessage.prototype, "label", void 0);
__decorate([
    class_validator_1.IsString(),
    class_validator_1.ValidateIf((o) => o.recipientKeys === undefined),
    __metadata("design:type", String)
], ConnectionInvitationMessage.prototype, "did", void 0);
__decorate([
    class_validator_1.IsString({
        each: true,
    }),
    class_validator_1.IsArray(),
    class_validator_1.ValidateIf((o) => o.did === undefined),
    class_validator_1.ArrayNotEmpty(),
    __metadata("design:type", Array)
], ConnectionInvitationMessage.prototype, "recipientKeys", void 0);
__decorate([
    class_validator_1.IsString(),
    class_validator_1.ValidateIf((o) => o.did === undefined),
    __metadata("design:type", String)
], ConnectionInvitationMessage.prototype, "serviceEndpoint", void 0);
__decorate([
    class_validator_1.IsString({
        each: true,
    }),
    class_validator_1.ValidateIf((o) => o.did === undefined),
    class_validator_1.IsOptional(),
    __metadata("design:type", Array)
], ConnectionInvitationMessage.prototype, "routingKeys", void 0);
__decorate([
    class_validator_1.IsOptional(),
    class_validator_1.IsUrl(),
    __metadata("design:type", String)
], ConnectionInvitationMessage.prototype, "imageUrl", void 0);
exports.ConnectionInvitationMessage = ConnectionInvitationMessage;
/**
 * Check whether an invitation is a `DIDInvitationData` object
 *
 * @param invitation invitation object
 */
function isDidInvitation(invitation) {
    return invitation.did !== undefined;
}
